engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.constraintvalidators.bv;
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import org.hibernate.validator.constraints.BitcoinAddress;
+import org.hibernate.validator.constraints.BitcoinAddressType;
+import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Checks that a given character sequence (e.g. string) is a well-formed BTC (Bitcoin) address.
+ *
+ * @author José Yoshiriro
+ */
+public class BitcoinAddressValidator implements ConstraintValidator<BitcoinAddress, CharSequence> {
+
+	private final List<BitcoinAddressType> addressType = new ArrayList<>();
+
+	private boolean singleType;
+	private String typesDescription;
+
+	@Override
+	public void initialize(BitcoinAddress bitcoinAddress) {
+		BitcoinAddressType[] types = bitcoinAddress.value();
+
+		this.singleType = ( types.length == 1 );
+
+		if ( Arrays.stream( types ).anyMatch( type -> type == BitcoinAddressType.ANY ) ) {
+			this.typesDescription = BitcoinAddressType.ANY.getDescription();
+			this.addressType.addAll(
+				Arrays.stream( BitcoinAddressType.values() )
+						.filter( type -> type != BitcoinAddressType.ANY )
+						.collect( Collectors.toList() ) );
+			return;
+		}
+
+		Collections.addAll( this.addressType, types );
+		this.typesDescription = this.addressType.stream()
+								.map( BitcoinAddressType::getDescription )
+								.collect( Collectors.joining( "; " ) );
```

Comment:
once the impl details are moved out of the enum here you could still keep them in the enum and maybe also have this prebuilt? e.g.:
```java
public enum BitcoinAddressConstraint {
	P2PKH("Legacy (P2PKH)", "^(1)[a-zA-HJ-NP-Z0-9]{25,61}$"),
	P2SH("Nested SegWit (P2SH)", "^(3)[a-zA-HJ-NP-Z0-9]{33}$"),
	BECH32("Native SegWit (Bech32)", "^(bc1)[a-zA-HJ-NP-Z0-9]{39,59}$"),
	P2WSH("SegWit variant of P2SH (P2WSH)", "^(bc1q)[a-zA-HJ-NP-Z0-9]{58}$"),
	P2WPKH("SegWit variant of P2PKH (P2WPKH)", "^(bc1q)[a-zA-HJ-NP-Z0-9]{38}$"),
	P2TR("Taproot (P2TR)", "^(bc1p)[a-zA-HJ-NP-Z0-9]{58}$");

	BitcoinAddressConstraint(String description, String pattern) {
		this.pattern = Pattern.compile( pattern );
	}

	private static final EnumMap<BitcoinAddress.Type, Collection<BitcoinAddressConstraint>> constraints;
	static {
		constraints = new EnumMap<>( BitcoinAddress.Type.class );
		constraints.put( BitcoinAddress.Type.ANY, Arrays.asList( BitcoinAddressConstraint.values() ) );
		constraints.put( BitcoinAddress.Type.P2PKH, Collections.singleton( P2PKH ) );
		constraints.put( BitcoinAddress.Type.P2SH, Collections.singleton( P2SH ) );
		constraints.put( BitcoinAddress.Type.BECH32, Collections.singleton( BECH32 ) );
		constraints.put( BitcoinAddress.Type.P2WSH, Collections.singleton( P2WSH ) );
		constraints.put( BitcoinAddress.Type.P2WPKH, Collections.singleton( P2WPKH ) );
		constraints.put( BitcoinAddress.Type.P2TR, Collections.singleton( P2TR ) );

	}
 ... 
}
```
and then in the initialize:
```java
public void initialize(BitcoinAddress bitcoinAddress) {
	Set<BitcoinAddressConstraint> constraints = new HashSet<>();
	for ( BitcoinAddress.Type type : bitcoinAddress.value() ) {
		constraints.addAll( BitcoinAddressConstraint.constraints.get( type ) );
	}
	this.singleType = constraints.size() == 1;
}
```

Reply:
I did almost that... When you have a chance to take a look please :)